src/codemodder/codetf/v3/codetf.py

```
@@ -0,0 +1,181 @@
+
+    @model_validator(mode="after")
+    def validate_description(self):
+        self.description = self.description or self.url
+        return self
+
+
+class Strategy(Enum):
+    ai = "ai"
+    hybrid = "hybrid"
+    deterministic = "deterministic"
+
+
+class AIMetadata(BaseModel):
+    provider: Optional[str] = None
+    models: Optional[list[str]] = None
+    total_tokens: Optional[int] = None
+    completion_tokens: Optional[int] = None
+    prompt_tokens: Optional[int] = None
+
+
+class GenerationMetadata(BaseModel):
+    strategy: Strategy
+    ai: Optional[AIMetadata] = None
+    provisional: bool
+
+
+class FixMetadata(BaseModel):
+    # Fix provider ID, corresponds to legacy codemod ID
+    id: str
+    # A brief summary of the fix
+    summary: str
+    # A detailed description of the fix
+    description: str
+    references: list[Reference]
+    generation: GenerationMetadata
+
+
+class Rating(BaseModel):
+    score: int
+    description: Optional[str] = None
+
+
+class FixQuality(BaseModel):
+    safetyRating: Rating
+    effectivenessRating: Rating
+    cleanlinessRating: Rating
+
+
+class FixResult(BaseModel):
+    """Result corresponding to a single finding"""
+
+    finding: Finding
+    fixStatus: FixStatus
+    changeSets: list[ChangeSet]
+    fixMetadata: Optional[FixMetadata] = None
+    fixQuality: Optional[FixQuality] = None
+    # A description of the reasoning process that led to the fix
+    reasoningSteps: Optional[list[str]] = None
+
+    @model_validator(mode="after")
+    def validate_fixMetadata(self):
+        if self.fixStatus.status == FixStatusType.fixed:
+            if not self.changeSets:
+                raise ValueError("changeSets must be provided for fixed results")
+            if not self.fixMetadata:
+                raise ValueError("fixMetadata must be provided for fixed results")
+        return self
+
+
+class CodeTF(CodeTFWriter, BaseModel):
+    run: Run
+    results: list[FixResult]
```

My expectation is that each finding in the input will have a corresponding result. I don't believe that's enforceable in the spec, and it probably should not be since we want to allow for the possibility of streaming subsets of results.

---

> I don't believe that's enforceable in the spec, and it probably should not be since we want to allow for the possibility of streaming subsets of results.

I'm not sure why these requirements are at odds. Aren't we only giving it 1 input as well? We had discussed as an input a single SARIF, and a single result ID (or something.) In this case, we'd expect 0 or 1 `FixResult`.

---

> Aren't we only giving it 1 input as well? We had discussed as an input a single SARIF, and a single result ID (or something.)

As a practical matter I'm still not convinced that's the best approach although we can work towards that setup. In any case I don't think we want to preclude the possibility of batch processing.

>  In this case, we'd expect 0 or 1 FixResult.

I would argue that the absence of any `FixResult` should be an exceptional condition. We should potentially consider making `FixResult` required. The fact that this would encode fix status means we should return a failure status if processing is failed or skipped